src/protocol/libp2p/kademlia/handle.rs

```diff
@@ -319,89 +320,25 @@ impl KademliaHandle {
         &mut self,
         key: RecordKey,
         public_addresses: Vec<Multiaddr>,
-    ) -> QueryId {
-        let query_id = self.next_query_id();
-        let _ = self
-            .cmd_tx
+    ) -> Result<QueryId, ()> {
+        let (query_id_tx, query_id_rx) = oneshot::channel();
+        self.cmd_tx
             .send(KademliaCommand::StartProviding {
                 key,
                 public_addresses,
-                query_id,
+                query_id_tx,
             })
-            .await;
+            .await
+            .map_err(|_| ())?;
 
-        query_id
+        query_id_rx.await.map_err(|_| ())
     }
 
     /// Store the record in the local store. Used in combination with
     /// [`IncomingRecordValidationMode::Manual`].
     pub async fn store_record(&mut self, record: Record) {
         let _ = self.cmd_tx.send(KademliaCommand::StoreRecord { record }).await;
     }
-
-    /// Try to add known peer and if the channel is clogged, return an error.
-    pub fn try_add_known_peer(&self, peer: PeerId, addresses: Vec<Multiaddr>) -> Result<(), ()> {
-        self.cmd_tx
-            .try_send(KademliaCommand::AddKnownPeer { peer, addresses })
-            .map_err(|_| ())
-    }
-
-    /// Try to initiate `FIND_NODE` query and if the channel is clogged, return an error.
-    pub fn try_find_node(&mut self, peer: PeerId) -> Result<QueryId, ()> {
-        let query_id = self.next_query_id();
-        self.cmd_tx
-            .try_send(KademliaCommand::FindNode { peer, query_id })
-            .map(|_| query_id)
-            .map_err(|_| ())
-    }
```

Comment:
Sad to see these removed. They were convenient in places where `.await` was not available, e.g., inside a `Stream::poll_next()` implementation and allowed an easy way to deal with backpressure. The alternative is of course to call `KademliaHandle::find_node()` and manually pin the future but it's not quite as ergonomic.

One thing worth pointing w.r.t. blocking handle -> protocol calls. I believe the program can deadlock if, e.g., `KademliaProtocol` is blocking on an `event_tx.send()` call (because the event channel is full) and if the command queue towards the protocol is full. Next call to, e.g., `KademliaHandle::find_node()` will also block and that could deadlock the program because both the handle and the protocol implementation are waiting on each other to drain the queues.

Reply:
Thanks for looking into this!

Basically, the alternative was to introduce either a mutex for generating the `QueryID`s, or introduce a compound
```
enum QueryId {
    User(UserQueryId),
    Local(LocalQueryId),
}
```
So we can generate `UserQueryId` in the handle and `LocalQueryId` in the protocol and they don't clash. But the implementation was too ugly with a lot of `match`es in places where such `QueryId` was used.

IDK, may be the plain mutex is a way to go.

Reply:
Request-response uses `Arc<AtomicUsize>` that's passed to the protocol via `RequestResponseConfig` to allocate `RequestId`s both in the handle and in the protocol implementation:

https://github.com/paritytech/litep2p/blob/e3a22d5cfbf255e74be1acceb3ea6ec0baa43c09/src/protocol/request_response/config.rs#L59

Would that work?

Reply:
Of course! Didn't think about atomics :see_no_evil:

Reply:
Should be resolved. Thanks again for review and suggestions!

Reply:
Thank you for bringing them back, much appreciated!